### PR TITLE
Add MainResponse version override as a cluster setting.

### LINKED ISSUE #693: [BUG] version number is incompatible with existing ES clients
**Describe the bug**
The API which provides version includes a field `version.number`. Clients parse this value to determine what version of ElasticSearch is running, and is used for checking API compatibility.

In ElasticSearch 7.10, this value is reported as `7.10`.

In OpenSearch 1.0.0, the value is `1.0.0-SNAPSHOT`.

Existing clients which parse this string are unable to communicate with OpenSearch when they rely on the value of the version.

**To Reproduce**

1. Start ElasticSearch 7.10
2. `curl http://localhost:9200`
3. Read `version.number`.
4. Start OpenSearch
5. `curl http://localhost:9200`
6. Read `version.number`.

**Expected behavior**

Because OpenSearch uses the same API as ElasticSearch 7.10.2, the value of `version.number` should be `7.10.2`.

Expected:

```
  "version" : {
    "number" : "7.10.2",
```

Actual:

```
  "version" : {
    "distribution" : "opensearch",
    "number" : "1.0.0-SNAPSHOT",
```

**Plugins**
N/A

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
N/A

**Additional context**

We confirmed that this is the cause of #667. It also appears that this may be causing #669.

### PR DESCRIPTION
**Pull request by mch2**
Signed-off-by: Marc Handalian <handalm@amazon.com>

### Description
This PR addresses #693 by using a cluster setting instead of a system property suggested by [814](https://github.com/opensearch-project/OpenSearch/pull/814)

This setting is intended to be set after a rolling upgrade via the cluster settings API.

```
PUT /_cluster/settings
{
  "compatibility" : {
    "override_main_response_version" : true
  }
}
``` 

It can be turned off with the same request with a value of false.

A default value for this setting may also be configured via opensearch.yml.  If not present, it defaults to false.

### Issues Resolved
#693 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
start gradle check

**Review by dblock — APPROVED** [review]
LGTM

**Comment by mch2** [comment]
Made an update here to improve the commit message

**Comment by CEHENKLE** [comment]
start gradle check

**Comment by mch2** [comment]
@nknize thoughts on this?

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java [hunk 1]
```diff
@@ -33,36 +33,63 @@
...
 
+    private static final DeprecationLogger DEPRECATION_LOGGER = DeprecationLogger.getLogger(TransportMainAction.class);
+
+    public static final String OVERRIDE_MAIN_RESPONSE_VERSION_KEY = "compatibility.override_main_response_version";
+
+    public static final Setting<Boolean> OVERRIDE_MAIN_RESPONSE_VERSION = Setting.boolSetting(
+        OVERRIDE_MAIN_RESPONSE_VERSION_KEY, false, Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    public static final String OVERRIDE_MAIN_RESPONSE_VERSION_DEPRECATION_MESSAGE = "overriding main response version" +
+        " number will be removed in version 2.0.0";
```
Comment (nknize) [inline server/src/main/java/org/opensearch/action/main/TransportMainAction.java:61]: One small suggestion: maybe we should reword this to `... will be removed in a future version`. Just in case this doesn't get addressed in 2.0.0? (alternatively let's open an issue to remove this, label it 2.0.0, and throw the issue link here so we commit to the removal in 2.0)

**Review by nknize — APPROVED** [review]
this looks great. I just threw one small suggestion so we don't lose the task of removing this in 2.0.0 (which is currently the main branch)

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java (hunk 1 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java]: Updated the wording to "in a future version".

**Comment by CEHENKLE** [comment]
start gradle check

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java (hunk 1 above)
Comment (nknize) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java]: :+1:

**Comment by gittygoo** [comment]
just tried using this to connect Kibana with opensearch and getting the following on kibana 

`This version of Kibana is incompatible with the following Elasticsearch nodes in your cluster:  v1.2.4 @ XXX.1.80:9200 (XXX.1.80), v1.2.4 @ XXX.2.133:9200 (XXX.2.133), v1.2.4 @ XXX.0.15:9200 (XXX.0.15)`

Note the version v1.2.4 is still being "seen" by kibana, wasnt it supposed to show something like 7.10.xxx ?

**Comment by dblock** [comment]
@gittygoo What version of Kibana are you trying to run?

Generally, I wouldn't ecommend running any version of Kibana on top of OpenSearch, even if we are able to change the version number returned here, this is an untested combination.

**Comment by dblock** [comment]
@gittygoo Reopen a new issue? I suspect that changing the version in nodes response is going to do other bad things :TM:
